protocols/perf/src/protocol.rs

```diff
@@ -18,184 +18,184 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use async_trait::async_trait;
-use futures::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use futures_timer::Delay;
 use instant::Instant;
-use libp2p_request_response as request_response;
-use libp2p_swarm::StreamProtocol;
-use std::io;
-
-use crate::{RunDuration, RunParams};
-
-const BUF: [u8; 65536] = [0; 64 << 10];
-
-#[derive(Debug)]
-pub enum Response {
-    Sender(usize),
-    Receiver(RunDuration),
-}
-
-#[derive(Default)]
-pub struct Codec {
-    to_receive: Option<usize>,
-
-    write_start: Option<Instant>,
-    read_start: Option<Instant>,
-    read_done: Option<Instant>,
+use std::time::Duration;
+
+use futures::{
+    future::{select, Either},
+    AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, FutureExt, SinkExt, Stream, StreamExt,
+};
+
+use crate::{Finished, Progressed, Run, RunDuration, RunParams, RunUpdate};
+
+const BUF: [u8; 1024] = [0; 1024];
+const REPORT_INTERVAL: Duration = Duration::from_secs(1);
+
+pub(crate) fn send_receive<S: AsyncRead + AsyncWrite + Unpin + Send + 'static>(
+    params: RunParams,
+    stream: S,
+) -> impl Stream<Item = Result<RunUpdate, std::io::Error>> {
+    // Use a channel to simulate a generator. `send_receive_inner` can `yield` events through the
+    // channel.
+    let (sender, receiver) = futures::channel::mpsc::channel(0);
+    let receiver = receiver.fuse();
+    let inner = send_receive_inner(params, stream, sender).fuse();
+
+    futures::stream::select(
+        receiver.map(|progressed| Ok(RunUpdate::Progressed(progressed))),
+        inner
+            .map(|finished| finished.map(RunUpdate::Finished))
+            .into_stream(),
+    )
```

Comment:
What about the following design:

1. Only use one channel.
2. Use `tokio::spawn` for `send_receive_inner`.
3. At the end of the loop, send a `RunUpdate::Finshed` into the channel, then drop it.
4. Return the receiver of the channel from `send_receive`, it automatically implements `Stream` for you.

Reply:
> Only use one channel.

This is already only using a single channel.

> Use `tokio::spawn` for `send_receive_inner`.

What would be the benefit of spawning an additional task?